src/pushed_authorization.rs

```diff
@@ -145,58 +112,100 @@ where
             user_hint,
             issuer_state,
         }
     }
 
-    pub fn request(
+    pub async fn async_request<C, F, RE>(
         self,
+        mut http_client: C,
         client_assertion_type: Option<String>,
         client_assertion: Option<String>,
-    ) -> Result<(String, serde_json::Value, CsrfToken), serde_json::Error> {
+    ) -> Result<(url::Url, CsrfToken), RequestError<RE>>
+    where
+        C: FnMut(HttpRequest) -> F,
+        F: Future<Output = Result<HttpResponse, RE>>,
+        RE: std::error::Error + 'static,
+    {
         let (url, token) = self.inner.url();
-        let mut body = json!({});
-        for (k, v) in url
-            .query_pairs()
-            .into_owned()
-            .collect::<HashMap<String, String>>()
-        {
-            body.as_object_mut().unwrap().entry(k).or_insert(json!(v));
-        }
-        body.as_object_mut()
-            .unwrap()
-            .entry("client_assertion_type")
-            .or_insert(json!(client_assertion_type));
-        body.as_object_mut()
-            .unwrap()
-            .entry("client_assertion")
-            .or_insert(json!(client_assertion));
-        body.as_object_mut()
+
+        let query = serde_urlencoded::from_str::<ParAuthParams>(url.clone().as_str())
             .unwrap()
-            .entry("authorization_details")
-            .or_insert(json!(&serde_json::to_string(&self.authorization_details)?));
-        if let Some(w) = self.wallet_issuer {
-            body.as_object_mut()
-                .unwrap()
-                .entry("wallet_issuer")
-                .or_insert(json!(&w.to_string()));
-        }
-        if let Some(h) = self.user_hint {
-            body.as_object_mut()
-                .unwrap()
-                .entry("user_hint")
-                .or_insert(json!(&h));
-        }
-        if let Some(s) = self.issuer_state {
-            body.as_object_mut()
-                .unwrap()
-                .entry("issuer_state")
-                .or_insert(json!(s.secret()));
+            .set_client_assertion_type(client_assertion_type.clone())
+            .set_client_assertion(client_assertion.clone())
+            .set_authorization_details(Some(
+                serde_json::to_string::<Vec<AuthorizationDetail<AD>>>(&self.authorization_details)
+                    .unwrap(),
+            ))
+            .set_wallet_issuer(self.wallet_issuer)
+            .set_user_hint(self.user_hint)
+            .set_issuer_state(self.issuer_state);
+
+        let http_request = HttpRequest {
+            url: self.par_auth_url.url().clone(),
+            method: Method::POST,
+            headers: vec![
+                (
+                    CONTENT_TYPE,
+                    HeaderValue::from_static(MIME_TYPE_FORM_URLENCODED),
+                ),
+                (ACCEPT, HeaderValue::from_static(MIME_TYPE_JSON)),
+            ]
+            .into_iter()
+            .collect(),
+            body: serde_json::to_vec(&format!(
```

Comment:
I think it's worth extracting this request building into a function and adding a unit test